src/libmeasurement_kit/neubot/dash_impl.hpp

```diff
@@ -0,0 +1,345 @@
+    size_t rate_index = 0;
+    while (dash_rates()[rate_index] < speed_kbit &&
+           rate_index < dash_rates().size()) {
+        rate_index++;
+    }
+    if (rate_index > 0) { // Make sure we don't underflow
+        rate_index -= 1;
+    }
+    return rate_index;
+}
+
+template <MK_MOCK_AS(http::request_send, http_request_send),
+          MK_MOCK_AS(http::request_recv_response, http_request_recv_response)>
+void run_loop_(Var<net::Transport> txp, int speed_kbit, std::string auth_token,
+               Var<report::Entry> entry, Settings settings,
+               Var<Reactor> reactor, Var<Logger> logger, Callback<Error> cb,
+               int iteration = 1) {
+
+    if (iteration > DASH_MAX_ITERATIONS) {
+        logger->debug("dash: completed all iterations");
+        cb(NoError());
+        return;
+    }
+
+    size_t rate_index = select_lower_rate_index(speed_kbit);
+    int rate_kbit = dash_rates()[rate_index];
+    int count = ((rate_kbit * 1000) / 8) * DASH_SECONDS;
+    std::string path = "/dash/download/";
+    path += std::to_string(count);
+    settings["http/path"] = path;
+    settings["http/method"] = "GET";
+    double saved_time = mk::time_now();
+    logger->debug("dash: requesting '%s'", path.c_str());
+
+    http_request_send(
+        txp, settings,
+        {
+            {"Authorization", auth_token},
+        },
+        "", [=](Error error, Var<http::Request> req) {
+            if (error) {
+                logger->warn("dash: request failed: %s",
+                             error.explain().c_str());
+                cb(error);
+                return;
+            }
+            assert(!!req);
+            http_request_recv_response(
+                txp,
+                [=](Error error, Var<http::Response> res) {
+                    if (error) {
+                        logger->warn("dash: cannot receive response: %s",
+                                     error.explain().c_str());
+                        cb(error);
+                        return;
+                    }
+                    assert(!!res);
+                    if (res->status_code != 200) {
+                        logger->warn("dash: invalid response code: %s",
+                                     error.explain().c_str());
+                        cb(http::HttpRequestFailedError());
+                        return;
+                    }
+                    res->request = req;
+                    double length = res->body.length();
+                    double time_elapsed = mk::time_now() - saved_time;
+                    if (time_elapsed <= 0) { // For robustness
+                        logger->warn("dash: invalid time error");
+                        cb(InvalidTimeError());
+                        return;
+                    }
+                    // TODO: we should fill all the required fields
+                    (*entry)["receiver_data"].push_back(report::Entry{
```

Comment:
I should probably write a OONI specification for this test saying exactly what should be saved by the OONI collector (the entry here should be what is saved by Neubot).